### DISCUSSION TITLE
Example of newer bitmaps

### DISCUSSION THREAD

**Discussion by larspalo** [GrandOrgue/grandorgue#1490] [Developers]
I've put together a few examples of different bitmaps ("moving" keys mostly) that still match the existing builtin bitmaps of GO. The small sizes severely limits the amount of details one can get, and the darker keys are very difficult to get really good results with. But anyway, here is my latest take on it! What do you think?

[TestNewGoImages.zip](https://github.com/GrandOrgue/grandorgue/files/11326380/TestNewGoImages.zip)

**Comment by midimusic** [comment]
GO keyboards are one of its weakest features. Particularly the fact that they cannot be "swiped" without staying down. Also in general they are too small to be of much value.

Having tried your new key designs I found that they are visually disappointing in particular it is almost impossible to see whether some of them (particularly sharps/flats) are up or down.

I think you need to decide whether the keys are just ornaments or whether they are to be functional. If ornaments then most existing organ designs are OK. If functional then they all leave a lot to be desired - including your new designs. Some of them being much worse than others.

Sorry to be a wet blanket but you did ask for opinions.

csw900

**Reply by larspalo** [reply]
No, hard feelings at all! It's good to hear your opinion.

Increasing size would allow for much greater details, but then it wouldn't be consistent with the already builtin bitmaps and changing them would break the expected layout. For a truly custom sample set it's of course better to use larger bitmaps, but my main idea was to increase the options for builtin bitmaps with a "moving" style that could be easily used without needing to specify individual on/off bitmaps for the keys of a manual.

My personal opinion of the displayed keyboards is that they are there for display and connectivity only, not for actual playing on! I think that for any serious usage GO should be played from MIDI keyboards, not from the screen.

What about changing the darker keys to actually lighten up when pressed instead of them going even darker as I've done now?

**Comment by larspalo** [comment]
Another version with a little increased range of motion and the dark bitmaps actually lighten slightly when depressed.

[TestNewGoImages2.zip](https://github.com/GrandOrgue/grandorgue/files/11333293/TestNewGoImages2.zip)

**Comment by oleg68** [comment]
I can not see which dark key is pressed, especially the sharp keys of the I and II manuals.

All other images look all right.

**Reply by larspalo** [reply]
@oleg68 Is that for both versions? Should I lighten up the pressed dark keys more do you think? Perhaps I should add a reversed color pedal version too?

**Reply by oleg68** [reply]
> s that for both versions?

I tested only the second one.

> Should I lighten up the pressed dark keys more do you think? 

I think it would be useful.

**Comment by midimusic** [comment]
You still need much more contrast between up and down keys. There is also a slightly corrupted display as the "black" images are bleeding through on to the lower parts of the "white" keys. You could consider giving the down keys a change of colour.

I do not think the manual/pedal displays nesessarily need to look anything like those of the original organ. Have a look at eplayOrgan's keys. If the display is to be useful it it very important to easily distinguish between up and down. I would use only white and black up images rather than using any of the darker ones. I would also make the pedal up images white and black.

csw900

**Reply by larspalo** [reply]
Well, it seems you've quite different goals than I do! Nothing wrong with that, but still it's definitely not what I search for now. I'm not looking for a high contrast key display suitable for the visually impaired at the moment.

My goal for this experiment is a set of moving keys that look as "authentic" as is possible within the imposed size limits, while still having on/off states that are relatively easily differed in between. Why? Because I'd like a quick and easy option to create a more "beautiful" keyboard style that possibly could be integrated in GO as an alternative to the red glowing borders of the current builtin bitmaps. For truly custom sample sets one can create bitmaps of any size one likes, but for now my intention is a set to possibly integrate with GO as an easy option instead of needing to specify individual bitmaps.

> If the display is to be useful it it very important to easily distinguish between up and down. I would use only white and black up images rather than using any of the darker ones. I would also make the pedal up images white and black.

Useful for what? Again, my target is definitely more for appearance than actual physical usage. Criteria are: providing options for connectivity (MIDI keyboards), compliance with the builtin bitmap sizes and "good looks". Choice of display is a matter of taste. To avoid the darker images one just need to avoid the DispKeyColourInverted=Y (default is anyway false) for the manual, but some others might want that style and your preferences should not be a limit for them. In my opinion, GO should (as it's currently doing, or at least tries to...) provide options that cater to different tastes.

**Comment by larspalo** [comment]
Yet a new version with increased movement also for the white keys, slightly more on/off differences for all keys, and finally also dark pedals that can be seen on it's own panel.

[TestNewGoImages3.zip](https://github.com/GrandOrgue/grandorgue/files/11342314/TestNewGoImages3.zip)

**Comment by midimusic** [comment]
Version 3 is definately an improvement. Unfortunately you did not originally state what your goals were. If these meet your goals then fine but I would avoid the top darkest set.

csw900

**Comment by oleg68** [comment]
@larspalo [TestNewGoImages3.zip](https://github.com/GrandOrgue/grandorgue/files/11342314/TestNewGoImages3.zip) - the sharp keys became better, but the becar keys on the third manual are still very dark when pressed.

**Comment by larspalo** [comment]
A new version where the 3rd manual black keys are even lighter when pressed, and the black pedal sharps have a little softer edges: 

[TestNewGoImages4.zip](https://github.com/GrandOrgue/grandorgue/files/11376270/TestNewGoImages4.zip)

**Comment by oleg68** [comment]
@larspalo the last images are all rights.

Are you going to contribute them to the GO source?

**Reply by larspalo** [reply]
> Are you going to contribute them to the GO source?

@oleg68 Only if you and @rousseldenis thinks it's worth including as it is right now. If we decide that we should not make any layout changes that will be backwards incompatible I think that this bitmap set might have a place in GO as improved built in "moving" bitmaps. On the other hand, if we decide that we can make changes in the layout (for instance built in bitmap sizes) for say a 4.0 version of GO that don't have to completely backwards compatible, then we could wait and create a completely new set with higher resolution.

I feel that this is about as good as I can get it within the current limitations. The real question is if we believe it's of any use for the GO community as it is. Personally, I think this set it better than the current options, but on the other hand, if I'd create a new custom sample set I'd most likely create completely new ones with higher resolution (different sizes). What this set of images would provide is just a quicker way to create "decent" moving bitmaps for the manuals for faster/quicker designs.

If we think this is "good enough" for the current situation, then the question arises if we should include it as an option or if it should replace the current bitmaps? As an option we could include a new boolean (for instance) DispMovingKeys that would default to false, but if true would replace current bitmaps with this set (including taking into account the inverted and wooden choices). What is your opinion?

**Reply by rousseldenis** [reply]
I've tested the last set of images and it looks great. All key changes are well visible. I like it!

> @oleg68 Only if you and @rousseldenis thinks it's worth including as it is right now. If we decide that we should not make any layout changes that will be backwards incompatible I think that this bitmap set might have a place in GO as improved built in "moving" bitmaps. On the other hand, if we decide that we can make changes in the layout (for instance built in bitmap sizes) for say a 4.0 version of GO that don't have to completely backwards compatible, then we could wait and create a completely new set with higher resolution.

I would say indeed this should be done in a major version of GO. That said, we should maybe begin a draft document to include ideas of great changes we want to see in GO, this included.

**Comment by oleg68** [comment]
@larspalo First of all, the new background and button images can be easy added to the current GO with next DisplayImageNum values.

I'd also make a similar capability to alter the manual and pedal images with DisplayImageNum. So they would be used only if the corresponding DisplayImageNum is specified in the ODF.

I don't think that it is a good practice to create several ODFs of the same sample set for different resolutions. I'd put a high-resolution layout only, because GrandOrgue can scale it to a lower one.

**Reply by larspalo** [reply]
> I don't think that it is a good practice to create several ODFs of the same sample set for different resolutions. I'd put a high-resolution layout only, because GrandOrgue can scale it to a lower one.

No, of course, as a sample set producer I'd work with the highest resolution I'd expect to be using. Now, the question is really if we want to continue using the current built in sizes and add this as an option, replace current, or wait and re-make all of the built in bitmaps to a new "scale" in a future version.

> I'd also make a similar capability to alter the manual and pedal images with DisplayImageNum. So they would be used only if the corresponding DisplayImageNum is specified in the ODF.

That would absolutely be an option. The current bitmaps could be kept as a default value and additional, new ones like this set, added beyond that. So do I understand it correctly that you would prefer adding DispImageNum to manual keys over a boolean moving value? That would allow even further additional styles while still keeping exact backwards compatibility!
